File: domain/src/main/java/com/catalog/domain/category/Category.java
package com.catalog.domain.category;

import com.catalog.domain.AggregateRoot;
import com.catalog.domain.utils.InstantUtils;
import com.catalog.domain.validation.ValidationHandler;

import java.time.Instant;
import java.util.Objects;

public class Category extends AggregateRoot<CategoryID> implements Cloneable {
    private String name;
    private String description;
    private Boolean active;
    private Instant createdAt;
    private Instant updatedAt;
    private Instant deletedAt;

    private Category(
            final CategoryID anId,
            final String aName,
            final String aDescription,
            final Boolean isActive,
            final Instant aCreatedAt,
            final Instant anUpdatedAt,
            final Instant aDeletedAt
    ) {
        super(anId);
        this.name = aName;
        this.description = aDescription;
        this.active = isActive;
        this.createdAt = Objects.requireNonNull(aCreatedAt, "'createdAt' should not be null");
        this.updatedAt = Objects.requireNonNull(anUpdatedAt, "'updatedAt' should not be null");
        this.deletedAt = aDeletedAt;
    }

    public static Category newCategory(final String aName, final String aDescription, final Boolean isActive) {
        final var id = CategoryID.unique();
        final var now = InstantUtils.now();
        final var deletedAt = isActive ? null : InstantUtils.now();

        return new Category(id, aName, aDescription, isActive, now, now, deletedAt);
    }

    public static Category with(
            final CategoryID anId,
            final String aName,
            final String aDescription,
            final Boolean isActive,
            final Instant aCreatedAt,
            final Instant anUpdatedAt,
            final Instant aDeletedAt
    ) {
        return new Category(anId, aName, aDescription, isActive, aCreatedAt, anUpdatedAt, aDeletedAt);
    }

    @Override
    public void validate(ValidationHandler handler) {
        new CategoryValidator(this, handler).validate();
    }

    public Category deactivate() {
        if  (getDeletedAt() == null) {
            this.deletedAt = InstantUtils.now();
        };

        this.active = false;
        this.updatedAt = InstantUtils.now();

        return this;
    }

    public Category activate() {
        this.active = true;
        this.updatedAt = InstantUtils.now();
        this.deletedAt = null;

        return this;
    }

    public Category update(
            final String aName,
            final String aDescription,
            final Boolean isActive
    ) {
        if (isActive) {
            activate();
        } else {
            deactivate();
        }

        this.name = aName;
        this.description = aDescription;
        this.updatedAt = InstantUtils.now();

        return this;
    }

    public CategoryID getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public String getDescription() {
        return description;
    }

    public Boolean isActive() {
        return active;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public Instant getUpdatedAt() {
        return updatedAt;
    }

    public Instant getDeletedAt() {
        return deletedAt;
    }

    @Override
    public Category clone() {
        try {
            Category clone = (Category) super.clone();
            // TODO: copy mutable state here, so the clone can't change the internals of the original
            return clone;
        } catch (CloneNotSupportedException e) {
            throw new AssertionError();
        }
    }
}

File: domain/src/main/kotlin/category/Category.kt
package category

import AggregateRoot
import validation.ValidationHandler
import java.time.Instant

class Category(
    anId: CategoryID,
    var name: String,
    var description: String?,
    var isActive: Boolean,
    val createdAt: Instant,
    var updatedAt: Instant,
    var deletedAt: Instant?
) : Cloneable, AggregateRoot<CategoryID>(anId) {
    companion object {
        fun newCategory(aName: String, aDescription: String?, isActive: Boolean): Category {
            val id = CategoryID.unique()
            val now = Instant.now()

            val deletedAt = if (isActive) null else now

            return Category(id, aName, aDescription, isActive, now, now, deletedAt)
        }
    }

    override fun validate(handler: ValidationHandler) {
        CategoryValidator(this, handler).validate()
    }

    fun deactivate(): Category {
        if (this.deletedAt == null) {
            this.deletedAt = Instant.now()
        }

        this.isActive = false
        this.updatedAt = Instant.now()

        return this
    }

    fun activate(): Category {
        this.deletedAt = null
        this.isActive = true
        this.updatedAt = Instant.now()

        return this
    }

    fun update(aName: String, aDescription: String, isActive: Boolean) : Category {
        if (isActive) {
            activate()
        } else {
            deactivate()
        }

        this.name = aName
        this.description = aDescription
        this.updatedAt = Instant.now()

        return this
    }

    public override fun clone(): Category {
        return super.clone() as Category
    }
}